feat: onchain payment method (#365)
cashubtc/nuts: 30.md

## Mint Settings

A `confirmations` option **SHOULD** be set to indicate the minimum depth in the blockchain for a transaction to be considered confirmed.

### Example `MintMethodSetting`

```json
{
  "method": "onchain",
  "unit": <str>,
  "min_amount": <int|null>,
  "max_amount": <int|null>,
  "options": {
    "confirmations": <int>
  }
}
```

## Melt Quote

For the `onchain` method, the wallet includes the following specific `PostMeltQuoteOnchainRequest` data:

```json
{
  "request": <str>,
  "unit": <str_enum[UNIT]>,
  "amount": <int>
}
```

Where:

- `request` is the Bitcoin address that should receive the onchain payment
- `unit` is the unit the wallet would like to pay with
- `amount` is the amount to send in the specified unit

Unlike other melt methods, a single `onchain` melt quote can contain multiple fee options for the same payment. This allows the wallet to choose between different fee and confirmation estimates while preserving a single quote ID.

The mint responds with a `PostMeltQuoteOnchainResponse`:

> ```json
> {
>   "quote": <str>,
>   "request": <str>,
>   "amount": <int>,
>   "unit": <str_enum[UNIT]>,
>   "fee_options": [
>     {
>       "fee": <int>,
>       "estimated_blocks": <int>
>     }
>   ],
>   "selected_estimated_blocks": <int|null>,
>   "state": <str_enum[STATE]>,
>   "expiry": <int>,
>   "outpoint": <str|null>
> }

Reviewer: nit - could we order the NUT-05 defaults first? makes it easier to see what has been added by the NUT

```suggestion
{
  "quote": <str>,
  "amount": <int>,
  "unit": <str_enum[UNIT]>,
  "state": <str_enum[STATE]>,
  "expiry": <int>,
  "request": <str>,
  "fee_options": [
    {
      "fee": <int>,
      "estimated_blocks": <int>
    }
  ],
  "selected_estimated_blocks": <int|null>,
  "outpoint": <str|null>
}
```

Author: addressed in 208799ca

# NUT-XX: Payment Method: Onchain

`optional`

`depends on: NUT-04 NUT-05 NUT-20`

---

This document describes minting and melting ecash with the `onchain` payment method, which uses Bitcoin onchain payments. It is an extension of [NUT-04][04] and [NUT-05][05] which cover the protocol steps of minting and melting ecash shared by any supported payment method.

## Mint Quote

For the `onchain` method, the wallet includes the following specific `PostMintQuoteOnchainRequest` data:

```json
{
  "unit": <str_enum[UNIT]>,
  "pubkey": <str>
}
```

> **Note:** A [NUT-20][20] `pubkey` is required in this NUT and the mint **MUST NOT** issue a mint quote if one is not included.

> **Privacy:** To prevent linking multiple mint quotes together, wallets **SHOULD** generate a unique public key for each mint quote request.

The mint responds with a `PostMintQuoteOnchainResponse`:

> ```json
> {
>   "quote": <str>,
>   "request": <str>,
>   "unit": <str_enum[UNIT]>,
>   "expiry": <int|null>,
>   "pubkey": <str>,
>   "amount_paid": <int>,
>   "amount_issued": <int>,
> }
> ```
> 
> Where:
> 
> - `quote` is the quote ID
> - `request` is the Bitcoin address to send funds to
> - `expiry` is the Unix timestamp until which the mint quote is valid
> - `pubkey` is the public key from the request
> - `amount_paid` is the total confirmed amount paid to the request
> - `amount_issued` is the amount of ecash that has been issued for the given mint quote

Reviewer 2: I think that is good to return the minimum confirmations from mint's to wallet. The wallet can check it outside before try to check the quote state.

````suggestion
```json
{
  "quote": <str>,
  "request": <str>,
  "unit": <str_enum[UNIT]>,
  "expiry": <int|null>,
  "pubkey": <str>,
  "min_confirmations": <int>,
  "amount_paid": <int>,
  "amount_issued": <int>,
}
```

Where:

- `quote` is the quote ID
- `request` is the Bitcoin address to send funds to
- `expiry` is the Unix timestamp until which the mint quote is valid
- `pubkey` is the public key from the request
- `min_confirmations` is the minimum number of bitcoin transaction confirmations specified for the mint's
- `amount_paid` is the total confirmed amount paid to the request
- `amount_issued` is the amount of ecash that has been issued for the given mint quote
````

Reviewer 2: OBS: That's already implemented on `MintMethodSetting`, but i think that's good to return on `PostMintQuoteOnchainResponse` too.

https://github.com/cashubtc/nuts/pull/365/changes#diff-f42c146fe5cf1807b0e4ef524f727dcfa12d8e2221480cf58577b5410d478c53R165

Author: I think including it in the quote implies it can be different per quote and is not a global config of the mint that should not change much so it can be kept in mint info.

### Example `MintMethodSetting`

```json
{
  "method": "onchain",
  "unit": <str>,
  "min_amount": <int|null>,
  "max_amount": <int|null>,
  "options": {
    "confirmations": <int>
  }
}
```

## Melt Quote

For the `onchain` method, the wallet includes the following specific `PostMeltQuoteOnchainRequest` data:

```json
{
  "request": <str>,
  "unit": <str_enum[UNIT]>,
  "amount": <int>
}
```

Where:

- `request` is the Bitcoin address that should receive the onchain payment
- `unit` is the unit the wallet would like to pay with
- `amount` is the amount to send in the specified unit

Unlike other melt methods, a single `onchain` melt quote can contain multiple fee options for the same payment. This allows the wallet to choose between different fee and confirmation estimates while preserving a single quote ID.

The mint responds with a `PostMeltQuoteOnchainResponse`:

```json
{
  "quote": <str>,
  "amount": <int>,
  "unit": <str_enum[UNIT]>,
  "state": <str_enum[STATE]>,
  "expiry": <int>,
  "request": <str>,

>   "fee_options": [
>     {
>       "fee_reserve": <int>,
>       "estimated_blocks": <int>
>     }

Reviewer 3: ```suggestion
    {
      "fee_index": <int>,
      "fee_reserve": <int>,
      "estimated_blocks": <int>
    }
```

  ],
  "selected_estimated_blocks": <int|null>,
  "outpoint": <str|null>
}
```

> Each item in `fee_options` represents one available fee reserve and confirmation estimate for the same payment. The wallet selects one of these options when executing the melt quote by including the option's `estimated_blocks` value in the melt request. The mint **MUST** return at least one `fee_options` item and **MUST NOT** return multiple `fee_options` items with the same `estimated_blocks` or `fee_reserve` value in one quote. The returned `fee_options` are fixed for the lifetime of the quote.

Reviewer 3: ```suggestion
Each item in `fee_options` represents one available fee reserve and confirmation estimate for the same payment. The wallet selects one of these options when executing the melt quote by including the option's `fee_index` value in the melt request. The mint **MUST** return at least one `fee_options` item. The returned `fee_options` are fixed for the lifetime of the quote.
```

> Where `fee_reserve` is the maximum onchain transaction fee the mint may charge for that option, and `estimated_blocks` is the estimated number of blocks until confirmation. `selected_estimated_blocks` is `null` before the quote is executed and is set to the selected `estimated_blocks` value once the wallet executes the quote. The mint expects the wallet to include `Proofs` of _at least_ `total_amount = amount + selected_fee_reserve + input_fee` where `selected_fee_reserve` is the `fee_reserve` from the selected `fee_options` item and `input_fee` is calculated from the keyset's `input_fee_ppk` as described in [NUT-02][02]. If the mint does not claim the full `selected_fee_reserve` as the actual fee, the mint returns the unclaimed amount as `change` to the wallet as described in [NUT-08][08].

Reviewer 3: ```suggestion
For each fee option with `fee_index`, `fee_reserve` is the maximum onchain transaction fee the mint may charge for that option, and `estimated_blocks` is the estimated number of blocks until confirmation. `selected_fee_index` is `null` before the quote is executed and is set by the mint to the selected fee option once the wallet executes the quote. The mint expects the wallet to include `Proofs` of _at least_ `total_amount = amount + selected_fee_reserve + input_fee` where `selected_fee_reserve` is the `fee_reserve` from the selected `fee_index` item and `input_fee` is calculated from the keyset's `input_fee_ppk` as described in [NUT-02][02]. If the mint does not claim the full `selected_fee_reserve` as the actual fee, the mint returns the unclaimed amount as `change` to the wallet as described in [NUT-08][08].
```

> Where `fee_reserve` is the maximum onchain transaction fee the mint may charge for that option, and `estimated_blocks` is the estimated number of blocks until confirmation. `selected_estimated_blocks` is `null` before the quote is executed and is set to the selected `estimated_blocks` value once the wallet executes the quote. The mint expects the wallet to include `Proofs` of _at least_ `total_amount = amount + selected_fee_reserve + input_fee` where `selected_fee_reserve` is the `fee_reserve` from the selected `fee_options` item and `input_fee` is calculated from the keyset's `input_fee_ppk` as described in [NUT-02][02]. If the mint does not claim the full `selected_fee_reserve` as the actual fee, the mint returns the unclaimed amount as `change` to the wallet as described in [NUT-08][08].

Reviewer 3: ```suggestion
Where `fee_reserve` is the maximum onchain transaction fee the mint may charge for that option, and `estimated_blocks` is the estimated number of blocks until confirmation. `selected_fee_option` is `null` before the quote is executed and is set to the selected `fee_index` value once the wallet executes the quote. The mint expects the wallet to include `Proofs` of _at least_ `total_amount = amount + selected_fee_reserve + input_fee` where `selected_fee_reserve` is the `fee_reserve` from the selected `fee_options` item and `input_fee` is calculated from the keyset's `input_fee_ppk` as described in [NUT-02][02]. If the mint does not claim the full `selected_fee_reserve` as the actual fee, the mint returns the unclaimed amount as `change` to the wallet as described in [NUT-08][08].
```

`state` is an enum string field with possible values `"UNPAID"`, `"PENDING"`, `"PAID"`:

- `"UNPAID"` means that the transaction has not been broadcast yet.
- `"PENDING"` means that the transaction is being processed by the mint but has not reached the required number of confirmations.
- `"PAID"` means that the transaction has been mined and confirmed.

`outpoint` is the transaction ID and output index of the payment in the format `txid:vout`, present once the transaction has been broadcast.

### Melting Tokens

Onchain melt requests are always asynchronous. The mint **MUST** return a `"PENDING"` state after validating the melt request and then broadcast the Bitcoin transaction in the background. The wallet **MUST** monitor the quote state through the check quote endpoint until the transaction reaches the required number of confirmations.

For the `onchain` method, the wallet includes the following specific `PostMeltOnchainRequest` data. The wallet can include an optional `outputs` field in the melt request to receive change for overpaid onchain fees (see [NUT-08][08]):

```json
{
  "quote": <str>,

>   "estimated_blocks": <int>,

Reviewer 3: ```suggestion
  "fee_index": <int>,
```

  "inputs": <Array[Proof]>,
  "outputs": <Array[BlindedMessage]> // Optional
}
```

> Where `estimated_blocks` is one of the values returned in the quote's `fee_options`. The mint **MUST** reject a melt request with an `estimated_blocks` value that was not returned in the quote. Once `selected_estimated_blocks` is set, the mint **MUST NOT** execute the quote again with a different `estimated_blocks` value.

Reviewer 3: ```suggestion
Where `fee_index` is the selected fee of the quote's `fee_options`. The mint **MUST** reject a melt request with a `fee_index` that was not returned in the quote. Once `selected_fee_index` is set, the mint **MUST NOT** execute the quote again with a different `fee_index` value.
```

If the `outputs` field is included and the mint does not claim the full `selected_fee_reserve` as the actual fee, the mint will respond with a `change` field containing blind signatures for the unclaimed amount (see [NUT-08][08]). The `change` field is omitted if there are no blind signatures to return. Note that because onchain transactions may be batched or have unpredictable costs, the mint is entitled to claim the full `selected_fee_reserve` as the actual fee.

```json
{
  "quote": <str>,
  "amount": <int>,
  "unit": <str_enum[UNIT]>,
  "state": <str_enum[STATE]>,
  "expiry": <int>,
  "request": <str>,

>   "fee_options": [
>     {
>       "fee_reserve": <int>,
>       "estimated_blocks": <int>
>     }

Reviewer 3: ```suggestion
    {
      "fee_index": <int>,
      "fee_reserve": <int>,
      "estimated_blocks": <int>
    }
```

  ],

>   "selected_estimated_blocks": <int>,

Reviewer 3: ```suggestion
  "selected_fee_index": <int>,
```

  "outpoint": <str|null>,
  "change": <Array[BlindSignature]> // Optional; present if outputs were included and there's change
}
```

### Example

**Melt quote request**:

```bash
curl -X POST https://mint.host:3338/v1/melt/quote/onchain -d \
'{"request": "bc1q...", "unit": "sat", "amount": 100000}'
```

**Melt quote response**:

```json
{
  "quote": "TRmjduhIsPxd...",
  "amount": 100000,
  "unit": "sat",
  "state": "UNPAID",
  "expiry": 1701704757,
  "request": "bc1q...",
  "fee_options": [
    {
      "fee_reserve": 5000,
      "estimated_blocks": 1
    },
    {
      "fee_reserve": 2000,
      "estimated_blocks": 6
    },
    {
      "fee_reserve": 800,
      "estimated_blocks": 144
    }
  ],

>   "selected_estimated_blocks": null,

Reviewer 3: ```suggestion
  "selected_fee_index": null,
```

>   "selected_estimated_blocks": null,

Reviewer 3: ```suggestion
  "selected_fee_index": null,
```

  "outpoint": null
}
```

Check quote state:

```bash
curl -X GET http://localhost:3338/v1/melt/quote/onchain/TRmjduhIsPxd...
```

**Melt request**:

```bash
curl -X POST https://mint.host:3338/v1/melt/onchain -d \
'{
  "quote": "TRmjduhIsPxd...",
  "estimated_blocks": 6,
  "inputs": [...],
  "outputs": [
    {
      "amount": 1,
      "id": "009a1f293253e41e",
      "B_": "03327fc4fa333909b70f08759e217ce5c94e6bf1fc2382562f3c560c5580fa69f4"
    }
  ]
}'
```

**Pending melt response**:

```json
{
  "quote": "TRmjduhIsPxd...",
  "amount": 100000,
  "unit": "sat",
  "state": "PENDING",

>   "expiry": 1701704757,
>   "request": "bc1q...",
>   "fee_options": [
>     {
>       "fee_reserve": 5000,
>       "estimated_blocks": 1
>     },
>     {
>       "fee_reserve": 2000,
>       "estimated_blocks": 6
>     },
>     {
>       "fee_reserve": 800,
>       "estimated_blocks": 144
>     }

Reviewer 3: ```suggestion
    {
      "fee_index": 0,
      "fee_reserve": 5000,
      "estimated_blocks": 1
    },
    {
      "fee_index": 1,    
      "fee_reserve": 2000,
      "estimated_blocks": 6
    },
    {
      "fee_index": 2,    
      "fee_reserve": 800,
      "estimated_blocks": 144
    }
```

  ],

>   "selected_estimated_blocks": 6,

Reviewer 3: ```suggestion
  "selected_fee_index": 1,
```

  "outpoint": null
}
```

> The wallet selects one of the returned `fee_options` by including that option's `estimated_blocks` value in the melt request. Once the quote is executed, quote state checks return the same response shape with `selected_estimated_blocks` set to the selected value and `outpoint` set once the transaction has been broadcast.

Reviewer 3: ```suggestion
The wallet selects one of the returned `fee_options` by including that option's `fee_index` value in the melt request. Once the quote is executed, quote state checks return the same response shape with `selected_fee_index` set to the selected value and `outpoint` set once the transaction has been broadcast.
```

**Paid quote state response**:

```json
{
  "quote": "TRmjduhIsPxd...",
  "amount": 100000,
  "unit": "sat",
  "state": "PAID",

>   "expiry": 1701704757,
>   "request": "bc1q...",
>   "fee_options": [
>     {
>       "fee_reserve": 5000,
>       "estimated_blocks": 1
>     },
>     {
>       "fee_reserve": 2000,
>       "estimated_blocks": 6
>     },
>     {
>       "fee_reserve": 800,
>       "estimated_blocks": 144
>     }

Reviewer 3: ```suggestion
    {
      "fee_index": 0,    
      "fee_reserve": 5000,
      "estimated_blocks": 1
    },
    {
      "fee_index": 1,    
      "fee_reserve": 2000,
      "estimated_blocks": 6
    },
    {
      "fee_index": 2,    
      "fee_reserve": 800,
      "estimated_blocks": 144
    }
```

  ],

>   "selected_estimated_blocks": 6,

Reviewer 3: ```suggestion
  "selected_fee_index": 1,
```
